File: coupling.m
function mtd = coupling(data,window)
%COUPLING       Multiplication of Temporal Derivatives
%
%  mtd = coupling(data,window);
%
%  This code creates a time-resolved connectivity matrix for a given window
%  length. See http://www.ncbi.nlm.nih.gov/pubmed/26231247 for more details.
%
%  Input:      data     time-series organized in 'time x nodes' matrix
%              window   window length (in time-points)
%
%  Output:     mtd      time-resolved connectivity matrix


%define variables
[ts,nodes] = size(data);
der = ts - 1;


%calculate temporal derivative
td = diff(data);


%standardize data
data_std = std(td);


for i = 1:nodes
    
    td(:,i) = td(:,i) / data_std(1,i);
    
end


% [...] = zscore(X,FLAG,DIM) standardizes X by working along the dimension
%    DIM of X. Pass in FLAG==0 to use the default normalization by N-1, or 1
%    to use N.


%functional coupling score

fc = bsxfun(@times,permute(td,[1,3,2]),permute(td,[1,2,3]));



%simple moving average

mtd_filter = 1/window*ones(window,1);

mtd = zeros(ts,nodes,nodes);

for j = 1:nodes
    for k = 1:nodes
        mtd(2:end,j,k) = filter(mtd_filter,1,fc(:,j,k));
    end
end

mtd(1:round(window/2),:,:) = [];
mtd(round(ts-window):ts,:,:) = 0;

mtd = permute(mtd,[2,3,1]);

mtd(:,:,1) = mtd(:,:,2);


end

File: coupling.py


import numpy as np
import pandas as pd

def coupling(data,window):
    """
        creates a functional coupling metric from 'data'
        data: should be organized in 'time x nodes' matrix
        smooth: smoothing parameter for dynamic coupling score
    """
    
    #define variables
    [tr,nodes] = data.shape
    der = tr-1
    td = np.zeros((der,nodes))
    td_std = np.zeros((der,nodes))
    data_std = np.zeros(nodes)
    mtd = np.zeros((der,nodes,nodes))
    sma = np.zeros((der,nodes*nodes))
    
    #calculate temporal derivative
    for i in range(0,nodes):
        for t in range(0,der):
            td[t,i] = data[t+1,i] - data[t,i]
    
    
    #standardize data
    for i in range(0,nodes):
        data_std[i] = np.std(td[:,i])
    
    td_std = td / data_std
   
   
    #functional coupling score
    for t in range(0,der):
        for i in range(0,nodes):
            for j in range(0,nodes):
                mtd[t,i,j] = td_std[t,i] * td_std[t,j]


    #temporal smoothing
    temp = np.reshape(mtd,[der,nodes*nodes])
    sma = pd.rolling_mean(temp,window)
    sma = np.reshape(sma,[der,nodes,nodes])
    
    return (fc, sma)
    
    

#input the variables 'd' (data) and 's' (smooth) 
#I've made 'd' random for now, but this could just as easily be real data
d = np.random.rand(200,5)
s = 9

#run the script
coupling(d,s)
